libr/util/json.c

```diff
@@ -0,0 +1,374 @@
+#define R_JSON_REPORT_ERROR(msg, p) fprintf(stderr, "R_JSON PARSE ERROR (%d): " msg " at %s\n", __LINE__, p)
+#else
+#define R_JSON_REPORT_ERROR(msg, p) do { (void)(msg); (void)(p); } while (0)
+#endif
+
+static RJson *json_new(void) {
+	return R_NEW0 (RJson);
+}
+
+static RJson *create_json(RJsonType type, const char *key, RJson *parent) {
+	RJson *js = json_new();
+	if (!js) {
+		return NULL;
+	}
+	js->type = type;
+	js->key = key;
+	if (!parent->children.last) {
+		parent->children.first = parent->children.last = js;
+	} else {
+		parent->children.last->next = js;
+		parent->children.last = js;
+	}
+	parent->children.length++;
+	return js;
+}
+
+R_API void r_json_free(RJson *js) {
+	if (!js) {
+		return;
+	}
+	if (js->type == R_JSON_OBJECT || js->type == R_JSON_ARRAY) {
+		RJson *p = js->children.first;
+		RJson *p1;
+		while (p) {
+			p1 = p->next;
+			r_json_free (p);
+			p = p1;
+		}
+	}
+	free (js);
+}
+
+static char *unescape_string(char *s, char **end) {
+	char *p = s;
+	char *d = s;
+	char c;
+	while ((c = *p++)) {
+		if (c == '"') {
+			*d = '\0';
+			*end = p;
+			return s;
+		} else if (c == '\\') {
+			switch (*p) {
+			case '\\':
+			case '/':
+			case '"':
+				*d++ = *p++;
+				break;
+			case 'b':
+				*d++ = '\b';
+				p++;
+				break;
+			case 'f':
+				*d++ = '\f';
+				p++;
+				break;
+			case 'n':
+				*d++ = '\n';
+				p++;
+				break;
+			case 'r':
+				*d++ = '\r';
+				p++;
+				break;
+			case 't':
+				*d++ = '\t';
+				p++;
+				break;
+			case 'u': { // unicode
+				char *ps = p - 1;
+				ut8 high = 0, low = 0;
+				if (r_hex_to_byte (&high, p[1]) || r_hex_to_byte (&high, p[2])
+						|| r_hex_to_byte (&low, p[3]) || r_hex_to_byte (&low, p[4])) {
+					R_JSON_REPORT_ERROR ("invalid unicode escape", p - 1);
+					return 0;
+				}
+				RRune codepoint = (RRune)high << 8 | (RRune)low;
+				if ((codepoint & 0xfc00) == 0xd800) { // high surrogate; need one more unicode to succeed
+					p += 6;
+					high = low = 0;
+					if (p[-1] != '\\' || *p != 'u'
+							|| r_hex_to_byte (&high, p[1]) || r_hex_to_byte (&high, p[2])
+							|| r_hex_to_byte (&low, p[3]) || r_hex_to_byte (&low, p[4])) {
+						R_JSON_REPORT_ERROR ("invalid unicode surrogate", ps);
+						return 0;
+					}
+					RRune codepoint2 = (RRune)high << 8 | (RRune)low;
+					if ((codepoint2 & 0xfc00) != 0xdc00) {
+						R_JSON_REPORT_ERROR ("invalid unicode surrogate", ps);
+						return 0;
+					}
+					codepoint = 0x10000 + ((codepoint - 0xd800) << 10) + (codepoint2 - 0xdc00);
+				}
+				int sz = r_utf8_encode ((ut8 *)d, codepoint);
+				if (!s) {
+					R_JSON_REPORT_ERROR ("invalid codepoint", ps);
+					return 0;
+				}
+				d += sz;
+				p += 5;
+				break;
+			}
+			default:
+				// leave untouched
+				*d++ = c;
+				break;
+			}
+		} else {
+			*d++ = c;
+		}
+	}
+	R_JSON_REPORT_ERROR ("no closing quote for string", s);
+	return 0;
+}
+
+static char *skip_block_comment(char *p) {
+	// assume p[-2]=='/' && p[-1]=='*'
+	char *ps = p - 2;
```

Comment:
This assumption is scary. It is properly checked on the caller? Or maybe it’s better to pass two pointers to this function to avoid this risk

Reply:
Changed to passing p - 2 directly.